readme.md
<!-- 
[//]: # (project_name = 'funTasia BloX World')
[//]: # (project_alias = 'funtasia')
[//]: # (file_name = readme)
[//]: # (file_extension = md)
[//]: # (file_format = markdown)
[//]: # (file_version = {major_version = 0, minor_version = 1, patch_version = 0, suffix_version = {'unreleased', 'alpha'}})
[//]: # [!] File version numbers have to match with that release version number where the file was last amended
[//]: # (author = {surname = 'Christian', lastname = 'Trant'})
[//]: # (date_of_creation = {day = 21, month = 05, year = 2017})
[//]: # (date_of_last_change = {day = 23, month = 05, year = 2017})
[//]: # (license = 'CC0') 
-->
# **funTasia BloX World** :smiley: :game_die: :earth_asia:

###### _A non-violent [Minetest](http://www.minetest.net/) subgame_

## Table of content

<!-- 
[//]: # (todo = 'set reference links')
-->
* [Preface](#preface)
* [What's new](#whats-new-in-this-release)
* [Features](#features)
* [Future prospects](#future-prospects)
* [Contribution](#contribution)
* [Licensing](#licensing)

## Preface

**funTasia BloX World** (this minetest subgame world module pack) was created for children who love to invent and create funTastic things rather than slay their pals or clean up their playgrounds.
<!-- 
[//]: # (todo = 'Place repo badges here')
[//]: # (---)
-->
[Minetest](http://www.minetest.net/) is a block game primarly inspired by [Minecraft](https://minecraft.net/de-de/). It is highly customizable by user scripts called mods. These mods are written in LUA script language and would be executed by an embedded [LUA](https://www.lua.org/) environment.

From the beginning of our childhood construction toys are present. They are considered as useful to train creativity and to learn how to abide rules. There are thousands of products out there - common toys and gaming software. Children start playing with new technologies at an increasingly early age. They explore the behavior of computer or smartphone apps and games intuitively without the ability to read. They are reliant on an intuitive and simple usablility. This would be arrived by a - non-cluttered - suitable pictograms, speech output and simple design instructions.  
Most of gaming software imply some violent characteristics. For example, the player have to kill some non-playing characters (NPC) to get the chance to a specific game item or in-game animals must be slaughtered for food by players themself.  
So children not only have to stress on the creative construction site, even more they must direct their focus to destructive details - at most in a matter of violence. Here *funTasia BloX World* comes into play.

Players in *funTasia BloX World* only have to focus on their
* social behavior in peer group 
* own creative skills

<!--
[//]: # (todo = 'add more fancy pros')
-->

Players in *funTasia BloX World* are prevented from
* destruct 'living environment' for subsistence
* fight nor destruct other players for benefits
* clean up their rooms and sandboxes
* violent speech

###### Information

I'm pretty new to game scripting. By this project, I want to improve my skills about LUA and performant game scripting.  
There are tons of awesome Minetest mods out there. At first I'll do some research about them to investigate *best practices*. Perhaps I am be able to collect them in a nice Wiki.  
If I did something wrong or made some misspellings, you are pleased to correct me because I'm not a native english speaker.  
Have I made some mistakes about licensing? Inform me in person and not by a lawyer. I'll declare a correction and a mention as soon as possible.

## What's new in this release

<!--
[//]: # (Excerpt of REALLY important changelog.md entries)
[//]: # (Stress to new features, improvements and major bugfixes)
-->

## Features

## Future prospects

- [ ] SBS - Simple Biomes System
  
  To keep the game as simple as possible for children, there should be no hard to reach and overburden areas. 
- [ ] ECS - Emoji Communication System
  - [ ] EmojiBloX - Emoji decoration blocks
  - [ ] EmojiSticKerZ - Emoji decoration
  - [ ] Unified EmojiBloX/EmojiSticKerZ configuration menu -> Administration
  - [ ] 
  
  I'm going to write an emoji system, which spawns emoji blocks over avatars head for some seconds. It should be possible to avoid text chat completely.
- NVS - Narrative Voice System
  
  All game information should presented by spoken word in the form of little narrative stories. Also information should explained by pictograms suitable for children and some kind of full functional, animated in-game tutorials.
- [ ] HS - Housing System
  How to realize that? There should be an own server for that. LOTRO has a nice housing system...
  - [ ] Idea: Housing for players (shared locked area)
  - [ ] Idea: Sandbox for players (shared locked area)
- [ ] WAS - World Administration System
  - [ ] UAS - User Administration System
- [ ] CYS - Construction Yard System
  
  Construction Yard System means, that there should be a creators version of funTasia. Just for setting up the whole ~~pixel~~ voxel world.
- [ ] KS - Karma System
  
  He who seeks his own happiness by not oppressing others, who also desire to have happiness, will find happiness in his next existence. ~ [Dhammapada Verse 132](http://www.tipitaka.net/tipitaka/dhp/verseload.php?verse=131)
  
  What the player gives should happen to the player thrice. So there a karma system should be implemented.
  - [ ] Idea: Ecological behavior - deforest gives one negative karma point
  - [ ] Idea: Ecological behavior - reforest gives one positive karma point
  - [ ] Idea: Ecological behavior - burn things gives one negative karma point
  - [ ] Idea: Ecological behavior - use renewable resources gives one positive karma point
  - [ ] Idea: Social behavior - Spend meal to other player gives one positive karma point
  - [ ] Idea: Social behavior - Communicate bad emoticons gives one negative karma point
  - [ ] Idea: Order behavior - Clean up without using a clean up tool gives two positive karma points
  - [ ] Idea: Order behavior - Put recipe to foliant, folder or book gives one positive karma point
  - a. s. o.
  
## Contribution

<!--
[//]: # (todo = 'add authors and licenses of used resources here')
[//]: # (todo = 'add link to contributing.md')
-->

## Licensing

![CC0](http://i.creativecommons.org/p/zero/1.0/88x31.png)

To the extent possible under law, Christian Trant has waived all copyright and related or neighboring rights to funTasia BloX World. This work is published from: Deutschland.

For further information see [license.md](https://github.com/FluttershyDeveloper/funTasia-BloX-World/blob/master/license.md) or http://creativecommons.org/publicdomain/zero/1.0/ .


changelog.md
<!--
[//]: # (project_name = 'funTasia BloX World')
[//]: # (project_alias = 'funtasia')
[//]: # (file_name = 'changelog')
[//]: # (file_extension = 'md')
[//]: # (file_format = 'markdown')
[//]: # (file_version = {major_version = 0, minor_version = 0, patch_version = 0, suffix_version = {'unreleased', 'alpha'}})
[//]: # [!] File version numbers have to match with that release version number where the file was last amended
[//]: # (author = {surname = 'Christian', lastname = 'Trant'})
[//]: # (date_of_creation = {day = 21, month = 05, year = 2017})
[//]: # (date_of_last_change = {day = 23, month = 05, year = 2017})
[//]: # (license = 'CC0')
-->
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

Possible entry types are 'Added', 'Changed', 'Depricated', 'Removed', 'Fixed' and 'Security'.

A release can be marked as 'YANKED'. Because of its security risks it mustn't be used anymore. Download must be prohibited.
## [0.2.0-unreleased-alpha] <!-- (YANKED) -->
## Added
- funtasia -> lua-module log_helper.lua added
- funtasia -> init.lua
  - tell_log_info(self, message) added
  - tell_error(self, message) added
  - tell_mod_identifier(self) added
  - some integrity checks added
## Changed
- funtasia -> init.lua refactored
  - local log_error(message) refactored to tell_log_error(self, message)
  - local log_message(log_type, message) refactored out to log_helper.write_log_message(log_type, mod_identifier, message)
### Depricated

### Removed

### Fixed

### Security

  ## [0.1.0-unreleased-alpha] <!-- (YANKED) -->

### Added
- emojiblox -> Emoji blocks
  - subset of EmojiOne emojis as blocks
  - descritption.txt added
- descritption.txt added
- init.lua
  - local log_error(message)
  - local log_message(log_type, message)
- commands.lua added

- recipes.lua added
  - Add cooking recipe: generic tree to 2x coal lump in 10 sec
  - Add cooking recipe: generic wood to coal lump in 8 sec
- CC0 Public Domain license added
### Changed
- init.lua refactored
  - loading recipes.lua
- debug logging improved 
  - log_info(message), log_error(message) -> local log_message(log_type, message)
  - set log_info(message) to local
### Depricated

### Removed

### Fixed

### Security

## [0.0.0-unreleased-alpha] - 2017-05-21 <!-- ([YANKED]) -->

## Added
- depends.txt
- init.lua
  - funtasia:get_version_string() added
  - log_info(message)
- license.md
  - [?] license model not sure yet
- contributing.md created
  - Olivier Lacan added
- changelog.md created
- readme.md created

# History of releases
[0.1.0-unreleased-alpha]: local
[0.0.0-unreleased-alpha]: local


recipes.lua
--# (project_name = 'funTasia BloX World')
--# (project_alias = 'funtasia')
--# (file_name = 'recipes')
--# (file_extension = 'lua')
--# (file_format = 'lua script')
--# (file_version = {major_version = 0, minor_version = 1, patch_version = 0, suffix_version = 'alpha'})
--# [!] File version numbers have to match with that release version number where the file was last amended
--# (author = {surname = 'Christian', lastname = 'Trant'})
--# (date_of_creation = {day = 21, month = 05, year = 2017})
--# (date_of_last_change = {day = 21, month = 05, year = 2017})
--# (license = 'CC0')

-- TODO: Refactor reciepes out to own module
-- Reciepes
--- Cooking - recipes for furnance

-- TODO: [?] Charcoal pile - Like creating a portal, the player must build a 'charcole pile' by a construction plan.
--- If 'charcoal pile' is constructed correctly, the player can light up a fuse and the wood inside the pile
--- transforms to charcoal blocks (low posibility) or charcoal lumps (height posibility).

-- TODO: Refactor to wood charcoal
-- add: charcoal_lump
-- color: 'dark_brown'
-- burntime: 26

-- add: charcoal_block
-- color: 'dark_brown'
-- burntime: 250

-- Bake tree to coal lump
minetest.register_craft({
  type = 'cooking',
  recipe = 'group:tree',
  output = 'default:coal_lump 2',
  cooktime = 10
})

-- Bake wood to coal lump
minetest.register_craft({
  type = 'cooking',
  recipe = 'group:wood',
  output = 'default:coal_lump',
  cooktime = 8
})


funtasia/log_helper.lua
--# (project_name = 'funTasia BloX World')
--# (project_alias = 'funtasia')
--# (file_name = 'log_helper')
--# (file_extension = 'lua')
--# (file_format = 'lua script')
--# (file_version = {major_version = 0, minor_version = 2, patch_version = 0, suffix_version = 'alpha'})
--# [!] File version numbers have to match with that release version number where the file was last amended
--# (author = {surname = 'Christian', lastname = 'Trant'})
--# (date_of_creation = {day = 04, month = 06, year = 2017})
--# (date_of_last_change = {day = 04, month = 06, year = 2017})
--# (license = 'CC0')

-- Define lua-module log_helper
local _log_helper = {}

  -- bool log_message(log_type, mod_identifier, message)
  -- Write a log 'message' of 'log_type' to debug.txt
  -- this is the abstraction for log_info(message) and log_error(message)
  local function write_log_message(log_type, mod_identifier, message)
    if not mod_identifier or not log_type or not message then return false end
    minetest.log(log_type, mod_identifier .. ' ' .. message)
    return true
  end
  
  -- bool log_info(message)
  -- Write a log message of type info to debug.txt
  -- Example: log_info('>TEST<')
  -- -> debug.txt -> 
  -- '[Mod] funTasia BloX World [0.1.0-unreleased-alpha] [funtasia] >TEST<' 
  function _log_helper.write_log_info(mod_identifier, message)
    local _result = false
    _result = write_log_message('info', mod_identifier, message)
    return _result
  end
				   
  -- bool log_error(message)
  -- Write a log message of type error to debug.txt
  function _log_helper.write_log_error(mod_identifier, message)
	local _result = false
	_result = write_log_message('error', mod_identifier, message)
	return _result
  end

return _log_helper

funtasia/init.lua
--# (project_name = 'funTasia BloX World')
--# (project_alias = 'funtasia')
--# (file_name = 'init')
--# (file_extension = 'lua')
--# (file_format = 'lua script')
--# (file_version = {major_version = 0, minor_version = 2, patch_version = 0, suffix_version = 'alpha'})
--# [!] File version numbers have to match with that release version number where the file was last amended
--# (author = {surname = 'Christian', lastname = 'Trant'})
--# (date_of_creation = {day = 21, month = 05, year = 2017})
--# (date_of_last_change = {day = 04, month = 06, year = 2017})
--# (license = 'CC0')
funtasia = {}

funtasia.name = minetest.get_current_modname()
funtasia.fullname = 'funTasia BloX World'
funtasia.path = minetest.get_modpath(funtasia.name)
funtasia.version = {major = 0,
                    minor = 2,
					patch = 0,
					suffixes = {'unreleased',
            					'alpha'}}
-- Initialization of the lua-module 'log_helper'
local _logger = dofile(funtasia.path .. '/log_helper.lua')

-- funtasia.tell_log_info(self, message)
-- tell the message as an info entry to debug.txt 
function funtasia.tell_log_info(self, message)
  local _result = false
  if type(message) ~= 'string' then 
    self:tell_error('message must be a string')
	return _result
  end
  _result = _logger.write_log_info(self.tell_mod_identifier(self), message)
  return _result
end

-- funtasia.tell_log_error(self, message)
-- tell the message as an error entry to debug.txt 
function funtasia.tell_log_error(self, message)
  local _result = false
  if type(message) ~= 'string' then 
    self:tell_error('message must be a string')
	return _result
  end
  _result = _logger.write_log_error(self.tell_mod_identifier(self), message)
  return _result
end

-- funtasia.tell_error(self, message)
-- tell the message as an error entry to debug.txt and cast it as a fatal error
function funtasia.tell_error(self, message)
  if type(message) ~= 'string' then 
    self:tell_error('message must be a string')
	return _result
  end
    local _message = 'ERROR:' .. self.tell_mod_identifier(self) .. ': ' .. tostring(message)
	self.tell_log_error(self, _message)
    error(_message)    
end

-- funtasia.get_version_string(self)
-- Get a version string with format 'major.minor.patch-suffixes'
function funtasia.tell_version_string(self)
  local result = 'M.m.p-(suffix)'
  if self.version.major and self.version.minor and self.version.patch then
    result = tostring(self.version.major) .. '.' ..
             tostring(self.version.minor) .. '.' ..
             tostring(self.version.patch)
  end
  if self.version.suffixes then
    local _suffixes = ''
	for _, entity in pairs(self.version.suffixes) do
	  _suffixes = _suffixes .. '-' .. tostring(entity)
	end
	result = result .. _suffixes
  end
  return result
end

-- funtasia.tell_mod_identifier(self)
-- Get a mod identifier string with format '[Mod] %self.fullname [%self.tell_version_string()] [%self.name]'
function funtasia.tell_mod_identifier(self)
  local _message = string.format('[Mod] %s [%s] [%s]',
                                 self.fullname,
	   		     				 self:tell_version_string(),
							     self.name)
  return _message
end

-- TODO: require libs?
dofile(funtasia.path .. '/commands.lua')
dofile(funtasia.path .. '/recipes.lua')
funtasia:tell_log_info('has been loaded.')
print('[Mod]' .. funtasia.fullname .. ' [' .. funtasia:tell_version_string() .. '] [' .. funtasia.name .. '] loaded.')